File: python/test_ApeTag_files.py
#!/usr/bin/env python
import ApeTag
import unittest
import os.path
import shutil
import filecmp

os.chdir('../test-files')

class TestApeTagFiles(unittest.TestCase):
    def _assertTagErrorRaised(self, msg, cb):
        try:
            cb()
            self.fail("No error raised")
        except ApeTag.TagError as e:
            self.assertEquals(e.message, msg)
        except UnicodeDecodeError:
            self.assertEquals(UnicodeDecodeError, msg)
        except AssertionError:
            raise
        except Exception as e:
            self.fail("non-TagError raised: " + str(e))

    def corrupt(self, filename, msg):
        self._assertTagErrorRaised(msg, lambda: ApeTag.getapefields(filename + '.tag'))

    def assertTagErrorRaised(self, msg, cb):
        self.withTestFile('good-empty', lambda: self._assertTagErrorRaised(msg, cb))

    def withTestFile(self, filename, cb):
        shutil.copy(filename + '.tag', 'test.tag')
        cb()
        os.remove('test.tag')
            
    def assertFilesEqual(self, before, after, cb):
        def f():
            cb()
            self.assertEquals(True, filecmp.cmp(after + '.tag', 'test.tag'))
        self.withTestFile(before, f)

    def assertRemoved(self, before, after):
        self.assertFilesEqual(before, after, lambda: ApeTag.deletetags('test.tag'))

    def test_corrupt(self):
        self.corrupt("corrupt-count-larger-than-possible", "Corrupt tag, end of tag reached with more items specified")
        self.corrupt("corrupt-count-mismatch", "Corrupt tag, mismatched header and footer item count")
        self.corrupt("corrupt-count-over-max-allowed", "Tag exceeds maximum allowed item count")
        self.corrupt("corrupt-data-remaining", "Corrupt tag, parsing complete but not at end of input: 545 bytes remaining")
        self.corrupt("corrupt-duplicate-item-key", "Corrupt tag, duplicate item key: 'name'")
        self.corrupt("corrupt-finished-without-parsing-all-items", "Corrupt tag, end of tag reached with more items specified")
        self.corrupt("corrupt-footer-flags", "Bad tag footer flags")
        self.corrupt("corrupt-header", "Nonexistent or corrupt tag, missing tag header")
        self.corrupt("corrupt-item-flags-invalid", "Corrupt tag, invalid item flags, bits 3-7 nonzero at position 32")
        self.corrupt("corrupt-key-invalid", "Corrupt tag, invalid item key at position 40: '\\x01ame'")
        self.corrupt("corrupt-key-too-short", "Corrupt tag, invalid item key at position 40: 'a'")
        self.corrupt("corrupt-key-too-long", "Corrupt tag, invalid item key at position 40: 'nnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn'")
        self.corrupt("corrupt-min-size", "Nonexistent or corrupt tag, missing tag header")
        self.corrupt("corrupt-missing-key-value-separator", "Corrupt tag, invalid item key at position 40: 'name\\x01valueAPETAGEX\\xd0\\x07'")
        self.corrupt("corrupt-next-start-too-large", "Corrupt tag, invalid item key at position 40: 'name\\x01valu'")
        self.corrupt("corrupt-size-larger-than-possible", "Existing tag says it is larger than the file: 65 bytes")
        self.corrupt("corrupt-size-mismatch", "Corrupt tag, header and footer sizes do not match")
        self.corrupt("corrupt-size-over-max-allowed", "Existing tag is too large: 61504 bytes")
        self.corrupt("corrupt-value-not-utf8", UnicodeDecodeError)
        self.corrupt("corrupt-item-length-invalid", UnicodeDecodeError)

    def test_has_ape(self):
        self.assertEquals(False, ApeTag.hasapetag("missing-ok.tag"))
        self.assertEquals(True, ApeTag.hasapetag("good-empty.tag"))
        self.assertEquals(False, ApeTag.hasapetag("good-empty-id3-only.tag"))
        self.assertEquals(True, ApeTag.hasapetag("good-empty-id3.tag"))

    def test_has_id3(self):
        self.assertEquals(False, ApeTag.hasid3tag("missing-ok.tag"))
        self.assertEquals(False, ApeTag.hasid3tag("good-empty.tag"))
        self.assertEquals(True, ApeTag.hasid3tag("good-empty-id3-only.tag"))
        self.assertEquals(True, ApeTag.hasid3tag("good-empty-id3.tag"))

    def test_parse(self):
        self.assertEquals({}, ApeTag.getapefields("good-empty.tag"))
        self.assertEquals({'name': ['value']}, ApeTag.getapefields("good-simple-1.tag"))
        #self.assertEquals(['value'], ApeTag.getapefields("good-simple-1.tag")['Name'])

        self.assertEquals(63, len(ApeTag.getapefields("good-many-items.tag")))
        self.assertEquals([''], ApeTag.getapefields("good-many-items.tag")['0n'])
        self.assertEquals(['a'], ApeTag.getapefields("good-many-items.tag")['1n'])
        self.assertEquals(['a' * 62], ApeTag.getapefields("good-many-items.tag")['62n'])

        self.assertEquals({'name': ['va', 'ue']}, ApeTag.getapefields("good-multiple-values.tag"))

        self.assertEquals('name', ApeTag.getapefields("good-simple-1.tag")['name'].key)
        self.assertEquals('utf8', ApeTag.getapefields("good-simple-1.tag")['name'].type)
        self.assertEquals(False, ApeTag.getapefields("good-simple-1.tag")['name'].readonly)
        
        self.assertEquals('name', ApeTag.getapefields("good-simple-1-ro-external.tag")['name'].key)
        self.assertEquals(['value'], ApeTag.getapefields("good-simple-1.tag")['name'])
        self.assertEquals('external', ApeTag.getapefields("good-simple-1-ro-external.tag")['name'].type)
        self.assertEquals(True, ApeTag.getapefields("good-simple-1-ro-external.tag")['name'].readonly)
        
        self.assertEquals('name', ApeTag.getapefields("good-binary-non-utf8-value.tag")['name'].key)
        self.assertEquals(['v\x81lue'], ApeTag.getapefields("good-binary-non-utf8-value.tag")['name'])
        self.assertEquals('binary', ApeTag.getapefields("good-binary-non-utf8-value.tag")['name'].type)
        self.assertEquals(False, ApeTag.getapefields("good-binary-non-utf8-value.tag")['name'].readonly)

        self.assertEquals({'name': ['value']}, ApeTag.getapefields(file("good-simple-1.tag")))

    def test_remove(self):
        self.assertRemoved('good-empty', 'missing-ok')
        self.assertRemoved('good-empty-id3', 'missing-ok')
        self.assertRemoved('good-empty-id3-only', 'missing-ok')
        self.assertRemoved('missing-10k', 'missing-10k')
        self.assertFilesEqual('good-empty', 'missing-ok', lambda: ApeTag.deletetags(file('test.tag', 'r+b')))

    def test_update(self):
        self.assertFilesEqual('good-empty', 'good-empty', lambda: ApeTag.createape('test.tag'))
        self.assertFilesEqual('missing-ok', 'good-empty', lambda: ApeTag.createape('test.tag'))
        self.assertFilesEqual('good-empty', 'good-simple-1', lambda: ApeTag.createape('test.tag', {'name': 'value'}))
        self.assertFilesEqual('good-simple-1', 'good-empty', lambda: ApeTag.updateape('test.tag', removefields=['name']))
        self.assertFilesEqual('good-simple-1', 'good-empty', lambda: ApeTag.updateape('test.tag', removefields=['Name']))
        self.assertFilesEqual('good-empty', 'good-simple-1-ro-external', lambda: ApeTag.createape('test.tag', {'name': ApeTag.ApeItem('name', ['value'], 'external', True)}))
        self.assertFilesEqual('good-empty', 'good-binary-non-utf8-value', lambda: ApeTag.createape('test.tag', {'name': ApeTag.ApeItem('name', ['v\x81lue'], 'binary')}))
        d = {}
        for i in range(63):
            d["%in" % i] = "a" * i
        self.assertFilesEqual('good-empty', 'good-many-items', lambda: ApeTag.createape('test.tag', d))
        self.assertFilesEqual('good-empty', 'good-multiple-values', lambda: ApeTag.createape('test.tag', {'name': ['va', 'ue']}))
        self.assertFilesEqual('good-multiple-values', 'good-simple-1-uc', lambda: ApeTag.createape('test.tag', {'NAME': 'value'}))
        self.assertFilesEqual('good-empty', 'good-simple-1-utf8', lambda: ApeTag.createape('test.tag', {'name': u'v\xc2\xd5'}))

        d['63n'] = 'a' * 63
        d['64n'] = 'a' * 64
        self.assertTagErrorRaised('New tag has too many items: 65 items', lambda: ApeTag.createape('test.tag', d))
        self.assertTagErrorRaised('New tag is too large: 8193 bytes', lambda: ApeTag.createape('test.tag', {'xn': 'a'*8118}))
        self.assertTagErrorRaised("Invalid item key for ape tag item: 'n'", lambda: ApeTag.createape('test.tag', {'n': 'a'}))
        self.assertTagErrorRaised("Invalid item key for ape tag item: 'n\\x00'", lambda: ApeTag.createape('test.tag', {'n\0': 'a'}))
        self.assertTagErrorRaised("Invalid item key for ape tag item: 'n\\x1f'", lambda: ApeTag.createape('test.tag', {'n\x1f': 'a'}))
        self.assertTagErrorRaised("Invalid item key for ape tag item: 'n\\x80'", lambda: ApeTag.createape('test.tag', {'n\x80': 'a'}))
        self.assertTagErrorRaised("Invalid item key for ape tag item: 'n\\xff'", lambda: ApeTag.createape('test.tag', {'n\xff': 'a'}))
        self.assertTagErrorRaised("Invalid item key for ape tag item: 'tag'", lambda: ApeTag.createape('test.tag', {'tag': 'a'}))
        self.assertTagErrorRaised(UnicodeDecodeError, lambda: ApeTag.createape('test.tag', {'ab': 'v\xff'}))
        self.assertTagErrorRaised("Invalid item type for ape tag item: 'foo'", lambda: ApeTag.createape('test.tag', {'ab': ApeTag.ApeItem('name', ['value'], 'foo')}))

        self.assertFilesEqual('good-empty', 'good-simple-1', lambda: ApeTag.createape(file('test.tag', 'r+b'), {'name': 'value'}))

    def test_id3(self):
        self.assertFilesEqual('missing-ok', 'good-empty-id3', lambda: ApeTag.createtags('test.tag'))
        self.assertFilesEqual('good-empty', 'good-empty-id3', lambda: ApeTag.createtags('test.tag'))
        self.assertFilesEqual('good-empty-id3', 'good-empty-id3', lambda: ApeTag.createtags('test.tag'))
        self.assertFilesEqual('good-empty-id3-only', 'good-empty-id3', lambda: ApeTag.createtags('test.tag'))

        self.assertFilesEqual('good-empty-id3', 'good-simple-4', lambda: ApeTag.createtags('test.tag', {'track':'1', 'genre':'Game', 'year':'1999', 'title':'Test Title', 'artist':'Test Artist', 'album':'Test Album', 'comment':'Test Comment'}))
        self.assertFilesEqual('good-empty-id3', 'good-simple-4-uc', lambda: ApeTag.createtags('test.tag', {'Track':'1', 'Genre':'Game', 'Year':'1999', 'Title':'Test Title', 'Artist':'Test Artist', 'Album':'Test Album', 'Comment':'Test Comment'}))
        self.assertFilesEqual('good-empty-id3', 'good-simple-4-date', lambda: ApeTag.createtags('test.tag', {'track':'1', 'genre':'Game', 'date':'12/31/1999', 'title':'Test Title', 'artist':'Test Artist', 'album':'Test Album', 'comment':'Test Comment'}))
        self.assertFilesEqual('good-empty-id3', 'good-simple-4-long', lambda: ApeTag.createtags('test.tag', {'track':'1', 'genre':'Game', 'year':'1999'*2, 'title':'Test Title'*5, 'artist':'Test Artist'*5, 'album':'Test Album'*5, 'comment':'Test Comment'*5}))
        
if __name__ == '__main__':
    unittest.main()

File: lua/test_apetag_files.lua
require('test_shared')

CHECK_ID3 = true

function tagname(file)
    return '../test-files/' .. file .. '.tag'
end

function tag(file)
    return ApeTag:new{filename=tagname(file), check_id3=CHECK_ID3}
end

function corrupt(file, msg)
    r, f = pcall(function() return tag(file):fields() end)
    assert(not r, 'No error raised, expected: ' .. msg)
    assert(string.find(f, msg, 1, true), 'Expected error: ' .. msg .. '\nReceived error: ' .. f)
    ASSERTIONS = ASSERTIONS + 2
end

function assert_has_field(file, field, values, flags)
    local i = tag(file):fields()[field]
    assert_equal(field, i.key)
    assert_equal(flags, i.flags)
    assert_arrays_equal(values, i)
end

function assert_files_equal(from, to, f)
    os.execute("cp " .. tagname(from) .. ' ' .. tagname('test'))
    f(tag('test'))
    local ret
    if _VERSION == 'Lua 5.1' then
      ret = 0
    else
      ret = true
    end
    assert_equal(ret, os.execute("cmp -s " .. tagname(to) .. ' ' .. tagname('test')))
    os.remove(tagname('test'))
end

function assert_update_raises(msg, f)
    os.execute("cp " .. tagname('good-empty') .. ' ' .. tagname('test'))
    r, errmsg = pcall(function() tag('test'):update(f) end)
    assert(not r, 'No error raised, expected: ' .. msg)
    assert(string.find(errmsg, msg, 1, true), 'Expected error: ' .. msg .. '\nReceived error: ' .. errmsg)
    ASSERTIONS = ASSERTIONS + 2
end

TestApeTag = {}

function TestApeTag.test_corrupt()
    corrupt("corrupt-count-larger-than-possible", "Item count is larger than possible")
    corrupt("corrupt-count-mismatch", "Header and footer item count does not match")
    corrupt("corrupt-count-over-max-allowed", "Item count is larger than than MAX_ITEM_COUNT")
    corrupt("corrupt-data-remaining", "Data remaining after specified number of items parsed")
    corrupt("corrupt-duplicate-item-key", "Multiple items with the same key")
    corrupt("corrupt-finished-without-parsing-all-items", "End of tag reached but more items specified")
    corrupt("corrupt-footer-flags", "Tag footer flags incorrect")
    corrupt("corrupt-header", "Missing header")
    corrupt("corrupt-item-flags-invalid", "Invalid item flags")
    corrupt("corrupt-item-length-invalid", "Invalid item length before taking key length into account")
    corrupt("corrupt-key-invalid", "Invalid ApeItem")
    corrupt("corrupt-key-too-short", "Invalid ApeItem")
    corrupt("corrupt-key-too-long", "Invalid ApeItem")
    corrupt("corrupt-min-size", "Tag size smaller than minimum size")
    corrupt("corrupt-next-start-too-large", "Invalid item length after taking key length into account")
    corrupt("corrupt-size-larger-than-possible", "Tag size larger than possible")
    corrupt("corrupt-size-mismatch", "Header and footer size does not match")
    corrupt("corrupt-size-over-max-allowed", "Tag size larger than possible")
    corrupt("corrupt-value-not-utf8", "Invalid ApeItem")
    corrupt("corrupt-missing-key-value-separator", "Missing key-value separator")
end

function TestApeTag.test_exists()
    assert_equal(tag('missing-ok'):exists(), false)
    assert_equal(tag('good-empty'):exists(), true)
    assert_equal(tag('good-empty-id3-only'):exists(), false)
    assert_equal(tag('good-empty-id3'):exists(), true)
end

function TestApeTag.test_fields()
    assert_tables_equal({}, tag('good-empty'):fields())
    assert_has_field('good-simple-1', 'name', {'value'}, 0)
    assert_tables_equal(tag('good-simple-1'):fields().Name, tag('good-simple-1'):fields().name)

    assert_has_field('good-many-items', '0n', {''}, 0)
    assert_has_field('good-many-items', '1n', {'a'}, 0)
    assert_has_field('good-many-items', '62n', {'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'}, 0)

    assert_has_field('good-multiple-values', 'name', {'va', 'ue'}, 0)
    assert_has_field('good-simple-1-ro-external', 'name', {'value'}, 5)
    assert_has_field('good-binary-non-utf8-value', 'name', {"v\129lue"}, 2)
end

function TestApeTag.test_remove()
    assert_files_equal('good-empty', 'missing-ok', function(tag) tag:remove() end)
    assert_files_equal('good-empty-id3', 'missing-ok', function(tag) tag:remove() end)
    assert_files_equal('good-empty-id3-only', 'missing-ok', function(tag) tag:remove() end)
    assert_files_equal('missing-10k', 'missing-10k', function(tag) tag:remove() end)
end

function TestApeTag.test_update()
    CHECK_ID3 = false
    assert_files_equal('good-empty', 'good-empty', function(tag) tag:update(function(fields) end) end)
    assert_files_equal('missing-ok', 'good-empty', function(tag) tag:update(function(fields) end) end)
    assert_files_equal('good-empty', 'good-simple-1', function(tag) tag:update(function(fields) fields.name = 'value' end) end)
    assert_files_equal('good-simple-1', 'good-empty', function(tag) tag:update(function(fields) fields.name = nil end) end)
    assert_files_equal('good-simple-1', 'good-empty', function(tag) tag:update(function(fields) fields.Name = nil end) end)
    assert_files_equal('good-empty', 'good-simple-1-ro-external', function(tag) tag:update(function(fields) fields.name = ApeItem:new('name', {'value'}, 5) end) end)
    assert_files_equal('good-empty', 'good-binary-non-utf8-value', function(tag) tag:update(function(fields) fields.name = ApeItem:new('name', {'v\129lue'}, 2) end) end)

    assert_files_equal('good-empty', 'good-many-items', function(tag) tag:update(function(fields)
      for i=0,62 do
        fields[i .. "n"] = string.rep("a", i)
      end
    end) end)

    assert_files_equal('good-empty', 'good-multiple-values', function(tag) tag:update(function(fields) fields.name = {'va', 'ue'} end) end)
    assert_files_equal('good-multiple-values', 'good-simple-1-uc', function(tag) tag:update(function(fields) fields.NAME = 'value' end) end)
    assert_files_equal('missing-ok', 'good-simple-1-utf8', function(tag) tag:update(function(fields) fields.name = 'v\195\130\195\149' end) end)

    assert_update_raises('Updated tag has too many items', function(fields)
      for i=0,64 do
        fields[i .. "n"] = string.rep("a", i)
      end
    end)
    assert_update_raises('Updated tag too large', function(fields) fields.xn = string.rep("a", 8118) end)
    assert_update_raises('Invalid ApeItem', function(fields) fields.n = 'a' end)
    assert_update_raises('Invalid ApeItem', function(fields) fields[string.rep("a", 256)] = 'a' end)
    assert_update_raises('Invalid ApeItem', function(fields) fields["v\0"] = 'a' end)
    assert_update_raises('Invalid ApeItem', function(fields) fields["v\31"] = 'a' end)
    assert_update_raises('Invalid ApeItem', function(fields) fields["v\129"] = 'a' end)
    assert_update_raises('Invalid ApeItem', function(fields) fields["v\255"] = 'a' end)
    assert_update_raises('Invalid ApeItem', function(fields) fields.tag = 'a' end)
    assert_update_raises('Invalid ApeItem', function(fields) fields.ab = 'v\129' end)
    assert_update_raises('Invalid ApeItem', function(fields) fields.name = ApeItem:new('name', {'value'}, 8) end)

    CHECK_ID3 = true

    assert_files_equal('missing-ok', 'good-empty-id3', function(tag) tag:update(function(fields) end) end)
    assert_files_equal('good-empty-id3-only', 'good-empty-id3', function(tag) tag:update(function(fields) end) end)

    assert_files_equal('good-empty-id3', 'good-simple-4', function(tag) tag:update(function(fields)
      fields.track = 1
      fields.genre = 'Game'
      fields.year = 1999
      fields.title = "Test Title"
      fields.artist = "Test Artist"
      fields.album = "Test Album"
      fields.comment = "Test Comment"
    end) end)

    assert_files_equal('good-empty-id3', 'good-simple-4-uc', function(tag) tag:update(function(fields)
      fields.Track = 1
      fields.Genre = 'Game'
      fields.Year = 1999
      fields.Title = "Test Title"
      fields.Artist = "Test Artist"
      fields.Album = "Test Album"
      fields.Comment = "Test Comment"
    end) end)

    assert_files_equal('good-empty-id3', 'good-simple-4-date', function(tag) tag:update(function(fields)
      fields.track = 1
      fields.genre = 'Game'
      fields.date = '12/31/1999'
      fields.title = "Test Title"
      fields.artist = "Test Artist"
      fields.album = "Test Album"
      fields.comment = "Test Comment"
    end) end)

    assert_files_equal('good-empty-id3', 'good-simple-4-long', function(tag) tag:update(function(fields)
      fields.track = 1
      fields.genre = 'Game'
      fields.year = 19991999
      fields.title = string.rep("Test Title", 5)
      fields.artist = string.rep("Test Artist", 5)
      fields.album = string.rep("Test Album", 5)
      fields.comment = string.rep("Test Comment", 5)
    end) end)

end

run_tests('test_apetag_files.lua')
